keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -261,6 +270,34 @@ details that didn't come across above? Go in to as much detail as
 necessary here. This might be a good place to talk about core concepts
 and how they relate. -->
 
+This is an important design note around selecting the group version for the new discovery types to be `meta/v1beta1`. [Link to the full comment](https://github.com/kubernetes/kubernetes/pull/111978#discussion_r979015557)
+
+1. Discovery is a non-resource API class
+2. As a non-resource API class, once the feature gate is
+   "on-by-default" the API is required to be stable (only additive
+   features)
+3. Non-resource APIs that are "off-by-default" do not promise
+   stability
+4. A non-resource APIs that has to change before promotion to
+   "on-by-default" must represent incompatible changes somehow to
+   clients (if the version is "v1" and then we find a bug, we would
+   have to rev to "v2" before "on-by-default", which means "v1" might
+   not ever be exposed to end users)
+5. Unversioned net new endpoints (/healthz) are effectively v1 even if
+   they are "off-by-default"
+6. We don't want to have multiple endpoints for discovery because it's
+   confusing for users and defeats the purpose of making discovery
+   more efficient, and we have a way to do that with negotiation
+7. We think there is value in a new API type (APIGroupDiscovery) which
+   simplifies client logic, but it comes with a small risk of not
+   being correct
+8. We have a good idea of what the API looks like due to a previous
+   v1, so we are evolving an existing API and are not "completely
+   flying blind" (i.e. implying this is really an alpha api)
+9. While we aren't exactly like an unversioned new endpoint (v1 from
+   start), we want to deliver the feature (improves clients) without
+   giving the perception that the API is perfect
+
```

Comment:
Also worth noting that this API is a GET, so the client doesn't provide a schema'd type, and this is not persisted in the cluster?

Reply:
I think this is covered by the first point that Discovery is a non-resource API class. Would you like it to be elaborated?

Reply:
> I think this is covered by the first point that Discovery is a non-resource API class. Would you like it to be elaborated?

we have resource APIs that are not persisted, but still have client provided schema. See subjectaccessreviews.  It even happens for other GET APIs, like pods/logs.

I think being explicit here will help future us.